File: day12/02.js
// const generations = 50000000000;
// const generations = 20;
const generations = 50000;
let count = -3;

const initialState =
  "...##.######...#.##.#...#...##.####..###.#.##.#.##...##..#...##.#..##....##...........#.#.#..###.#";

const map = {
  ".###.": "#",
  "#.##.": ".",
  ".#.##": "#",
  "...##": ".",
  "###.#": "#",
  "##.##": ".",
  ".....": ".",
  "#..#.": "#",
  "..#..": "#",
  "#.###": "#",
  "##.#.": ".",
  "..#.#": "#",
  "#.#.#": "#",
  ".##.#": "#",
  ".#..#": "#",
  "#..##": "#",
  "##..#": "#",
  "#...#": ".",
  "...#.": "#",
  "#####": ".",
  "###..": "#",
  "#.#..": ".",
  "....#": ".",
  ".####": "#",
  "..###": ".",
  "..##.": "#",
  ".##..": ".",
  "#....": ".",
  "####.": "#",
  ".#.#.": ".",
  ".#...": "#",
  "##...": "#"
};

const mem = {};

calculate();

function calculate() {
  const head = "..";
  let state = initialState;
  for (let i = 0; i < generations; i++) {
    state = generateNewState(head + state);
    if (mem[state]) {
      state = mem[state];
    } else {
      // console.log(i, state.length);
      // state = head + state;
      // state = generateNewState(state);
      // console.log(state);
      // console.log(count);
      const newState = cal(state);
      mem[state] = newState;
      state = newState;
    }
  }
  const ans = calculateStateValue(state);
  console.log(Object.keys(mem).length);
  console.log(ans);
}

function calculateStateValue(state) {
  let sum = 0;
  let start = count;

  const arr = state.split("");

  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === "#") {
      sum += start;
    }
    start++;
  }
  return sum;
}

function cal(state) {
  let str = "";
  let st = state;

  while (true) {
    if (st === ".....") {
      str += ".";
      break;
    }
    if (st.length < 5) {
      st += ".";
      continue;
    }

    const pattern = st.slice(0, 5);
    const s = map[pattern] || ".";
    st = st.slice(1);
    str += s;
  }

  return str;
}

function generateNewState(state) {
  let st = state;

  // console.log("=========");
  // console.log(st);
  while (true) {
    const pattern = st.slice(0, 5);
    // console.log(pattern);
    if (pattern !== ".....") {
      return st;
    }
    count++;
    st = st.slice(1);
  }
}

File: day12/02.ex
defmodule Logic do
  def find(state, transform_map) do
    head = ".."

    ans =
      1..10000
      |> Enum.reduce(
        %{
          state: state,
          count: -3
        },
        fn _g, data ->
          st = data.state
          count = data.count
          {gen_state, new_count} = generate_new_state(head <> st, count)
          # IO.puts(st)
          new_state = cal(gen_state, transform_map)

          %{
            state: new_state,
            count: new_count
          }
        end
      )

    # |> IO.inspect()
    sum(ans.state, ans.count)
    # |> IO.inspect()
  end

  def generate_new_state(state, count) do
    pattern = String.slice(state, 0..4)

    if pattern == "....." do
      generate_new_state(String.slice(state, 1..-1), count + 1)
    else
      {state, count}
    end
  end

  def sum(state, start) do
    state
    |> String.graphemes()
    |> Enum.reduce(
      %{
        sum: 0,
        count: start
      },
      fn s, data ->
        if s == "#" do
          new_sum = data.sum + data.count
          new_count = data.count + 1

          %{
            count: new_count,
            sum: new_sum
          }
        else
          %{
            count: data.count + 1,
            sum: data.sum
          }
        end
      end
    )
  end

  def cal(state, map) do
    # IO.puts(state)

    if state == "....." do
      "."
    else
      if String.length(state) < 5 do
        cal(state <> ".", map)
      else
        pattern = String.slice(state, 0..4)

        # if Map.has_key?(map, pattern) do
        #   IO.puts(pattern)
        # end

        # IO.puts(Map.has_key?(map, pattern))
        # mid = String.at(pattern, 2)

        new_mid = Map.get(map, pattern, ".")
        new_mid <> cal(String.slice(state, 1..-1), map)
      end
    end
  end
end

initialState =
  "..." <>
    "##.######...#.##.#...#...##.####..###.#.##.#.##...##..#...##.#..##....##...........#.#.#..###.#"

map = %{
  ".###.": "#",
  "#.##.": ".",
  ".#.##": "#",
  "...##": ".",
  "###.#": "#",
  "##.##": ".",
  ".....": ".",
  "#..#.": "#",
  "..#..": "#",
  "#.###": "#",
  "##.#.": ".",
  "..#.#": "#",
  "#.#.#": "#",
  ".##.#": "#",
  ".#..#": "#",
  "#..##": "#",
  "##..#": "#",
  "#...#": ".",
  "...#.": "#",
  "#####": ".",
  "###..": "#",
  "#.#..": ".",
  "....#": ".",
  ".####": "#",
  "..###": ".",
  "..##.": "#",
  ".##..": ".",
  "#....": ".",
  "####.": "#",
  ".#.#.": ".",
  ".#...": "#",
  "##...": "#"
}

# initialState = "..." <> "#..#.#..##......###...###"

# map = %{
#   "...##": "#",
#   "..#..": "#",
#   ".#...": "#",
#   ".#.#.": "#",
#   ".#.##": "#",
#   ".##..": "#",
#   ".####": "#",
#   "#.#.#": "#",
#   "#.###": "#",
#   "##.#.": "#",
#   "##.##": "#",
#   "###..": "#",
#   "###.#": "#",
#   "####.": "#"
# }

str_key_map =
  map
  |> Map.keys()
  |> Enum.reduce(%{}, fn k, m ->
    k_str = Atom.to_string(k)
    Map.put(m, k_str, Map.get(map, k))
  end)

IO.inspect(str_key_map)

Logic.find(initialState, str_key_map)
|> IO.inspect()
